node/lib/util/checkout.js

```diff
@@ -62,16 +64,45 @@ exports.checkout = co.wrap(function *(metaRepo, committish) {
     }
     const commit = yield metaRepo.getCommit(annotated.id());
 
-    yield GitUtil.setHeadHard(metaRepo, commit);
 
     const open = yield SubmoduleUtil.listOpenSubmodules(metaRepo);
     const names = yield SubmoduleUtil.getSubmoduleNamesForCommit(metaRepo,
                                                                  commit);
     const shas = yield SubmoduleUtil.getSubmoduleShasForCommit(metaRepo,
                                                                names,
                                                                commit);
+    // First, do dry runs.
 
-    // checkout submodules
+    let errors = [];
+
+    /**
+     * If it is possible to check out the specified `commit` in the specified
+     * `repo`, return `null`; otherwise, return an error message.
+     */
+    const dryRun = co.wrap(function *(repo, commit) {
+        try {
+            yield NodeGit.Checkout.tree(repo, commit, {
+                checkoutStrategy: NodeGit.Checkout.STRATEGY.NONE,
+            });
+            return null;                                              // RETURN
+        }
+        catch(e) {
+            return e.message;                                         // RETURN
+        }
+    });
+
+    // Check meta
+
+    const metaError = yield dryRun(metaRepo, commit);
+    if (null !== metaError) {
+        errors.push(`Unable to check out meta-repo: ${metaError}.`);
+    }
+
+    // Try the submodules; store the opened repos and loaded commits for use
+    // in the actual checkout later.
+
+    let subRepos = [];     // will contain a list of repositories for sub-repos
+    let subCommits = [];   // will contain a list of commits to checkout
 
     yield open.map(co.wrap(function *(name) {
         // Open repo but not alive on this commit.
```

Comment:
How do we actually want to handle this case?  This is, in effect, checking out a deletion, right?